src/content/post/stacksafe-taming-recursive-algorithms-in-rust-without-stack-overflow.mdx

```diff
@@ -0,0 +1,272 @@
+    if n <= 1 { 1 } else { n * factorial(n - 1) }
+}
+
+// No panic, works perfectly!
+println!("{}", factorial(100000));
+```
+
+`StackSafe` is being used in production by products like [ScopeDB](https://www.scopedb.io/blog/manage-observability-data-in-petabytes), where it helps trace and debug petabyte-scale observability data workloads.
+
+## The Stack Overflow Problem
+
+Recursive algorithms are elegant and intuitive, but they come with a fundamental limitation: stack overflow. Consider these common scenarios:
+
+```rust
+// Deep tree traversal - will crash on large trees
+fn tree_sum(node: &TreeNode) -> i32 {
+    match node {
+        TreeNode::Leaf(value) => *value,
+        TreeNode::Branch(left, right) => tree_sum(left) + tree_sum(right),
+    }
+}
+
+// JSON parsing - will crash on deeply nested JSON
+fn parse_value(tokens: &mut TokenStream) -> JsonValue {
+    match tokens.peek() {
+        Token::LeftBrace => parse_object(tokens),  // Recursive call
+        Token::LeftBracket => parse_array(tokens), // Recursive call
+        _ => parse_primitive(tokens),
+    }
+}
+```
+
+A sufficiently deep structure will cause your program to crash with `stack overflow`, and there's no clean way to predict or handle this in standard Rust.
+
+## Existing Solutions
+
+### Manual Transformation to Iterative Code
+
+The most common approach is rewriting recursive algorithms as loops with explicit stacks:
+
+```rust
+fn tree_sum_iterative(root: &TreeNode) -> i32 {
+    let mut stack = vec![root];
+    let mut sum = 0;
+    while let Some(node) = stack.pop() {
+        match node {
+            TreeNode::Leaf(value) => sum += value,
+            TreeNode::Branch(left, right) => {
+                stack.push(left);
+                stack.push(right);
+            }
+        }
+    }
+    sum
+}
+```
+
+This approach works for simple cases but becomes extremely complex or impossible when:
+
+1. The algorithm transforms data structures rather than just evaluating them (e.g., optimizing an AST)
+2. Multiple recursive calls need to be coordinated (e.g., tree balancing algorithms)
+3. The algorithm doesn't fit the tail-recursion pattern
+
+### Lower-Level Crates: `stacker` and `recursive`
+
+- [stacker](https://crates.io/crates/stacker): Provides low-level stack growth mechanisms
+- [recursive](https://crates.io/crates/recursive): Provides a `#[recursive]` to ease the application of `stacker`
+
+**Limitations**:
+
+- You must carefully not leaving any recursive functions not annotated with `#[recursive]`
+- Derived traits like `Debug`, `Clone`, and `Drop` on deeply nested structures still cause stack overflow, you must manually implement all traits with stack protection:
+
+```rust
+#[derive(Clone, Debug)]  // These will still stack overflow!
+enum Tree {
+    Leaf(i32),
+    Node(Box<Tree>, Box<Tree>),
+}
+
+let deep_tree = create_deep_tree(10000);
+let cloned = deep_tree.clone();  // Stack overflow here
+println!("{:?}", cloned);        // Or here
```

Comment:
This example is not quite clear. I wonder if we can host full examples in stacksafe repo, or even run them as tests.